feat(blog): kefka post (#1231)
Xe/site: lume/src/blog/2026/dancing-mad-sandboxing.mdx

<ConvP>
  <Conv name="Aoi" mood="wut">
    If I recall, some POSIX utilities like `false` aren't usable as Go identifiers, how did you handle package names for that?
  </Conv>
  <Conv name="Cadey" mood="aha">
    By naming them things like `falsecmd`:

    ```go
    package falsecmd

    // ...

    type Impl struct{}

    func (Impl) Exec(ctx context.Context, ec *command.ExecContext, args []string) error {
    	return interp.ExitStatus(1)
    }
    ```

    Honestly the implementations of `true` and `false` are my favourite part of this implementation. Here's the implementation of `true`:

    ```go
    package truecmd

    // ...

    type Impl struct{}

    func (Impl) Exec(ctx context.Context, ec *command.ExecContext, args []string) error {
    	return nil
    }
    ```

    This is a fully POSIX compliant implementation of `true`! Here's the relevant part of the spec if you don't believe me:

    > `true` - return true value
    >
    > **SYNOPSIS**
    >
    > `true`
    >
    > **OPTIONS**
    >
    > None.
    >
    > **OPERANDS**
    >
    > None.

>     I would link to the exact page in the POSIX spec, but they use a legacy frameset layout so you have to [dig through the Shell and Utilities section](https://pubs.opengroup.org/onlinepubs/9799919799/).

Author: Link to https://pubs.opengroup.org/onlinepubs/9799919799/utilities/true.html

  </Conv>
</ConvP>

### I made an operating system\*

> This is basically an operating system: it provides interfaces for programs (functions) to get input from a user, send output to a user, interact with a filesystem, and more. Eventually I want to add networking — a network stack on ExecContext, probably tsnet or wireguard-go for the user-level side.

Author: ```suggestion
This is basically an operating system: it provides interfaces for programs (well, in this case functions) to get input from a user, send output to a user, interact with a filesystem, and more. Eventually I want to add networking via a network stack on ExecContext, probably with [tsnet](https://pkg.go.dev/tailscale.com/tsnet) or [wireguard-go's netstack package](https://pkg.go.dev/golang.zx2c4.com/wireguard/tun/netstack) for the user-level side. Maybe there's room for adding [CEL](https://cel.dev/) based network filters there too.
```

## Porting applications with WebAssembly

Once I got basic coreutils working, I thought it would be fun to get Python, jq, and ripgrep working. From [previous experimentation](https://tangled.org/xeiaso.net/x/tree/master/llm/codeinterpreter/python) [back in the strawberry era of AI](https://xeiaso.net/blog/2024/strawberry/), I had already gotten Python running in WebAssembly via [wazero](https://github.com/wazero/wazero). This used the stdlib [io/fs#FS](https://pkg.go.dev/io/fs#FS) interface to allow me to inject virtual filesystems into the WebAssembly context. I used this to isolate my chatbot's filesystem state so that it (hopefully) wasn't able to delete anything important by accident.

> `io/fs#FS` has methods for the important stuff, and runtime interface assertions let you bridge the gap for things like writes. But it was really designed for [embedded filesystems](https://pkg.go.dev/embed#hdr-File_Systems), and writes get hairy fast once you're talking to object storage or anything that isn't a tree of bytes on disk.

Author: ```suggestion
`io/fs#FS` has methods for the important stuff, and runtime interface assertions let you bridge the gap for things like writes. But it was really designed for [embedded filesystems](https://pkg.go.dev/embed#hdr-File_Systems), and writes get hairy fast once you're talking to object storage or anything that isn't a tree of bytes on disk. This was unfortunate but understandable, io/fs#FS was initially made for embedding static data into Go programs and it shows.
```

At some point I hit a wall and had to switch from io/fs#FS to [billy](https://github.com/go-git/go-billy), another filesystem interface that I think predates the standard library one. This gives you [a bunch more methods](https://pkg.go.dev/github.com/go-git/go-billy/v5@v5.9.0#Filesystem) that map a lot closer to filesystem semantics in ways that coreutils crave. The interface was also mostly compatible with io/fs#FS so most of the hard part was really changing out the type and then chasing down compiler errors until I found enough of a pattern to have Opus automate the rest of it.

> From there it was a matter of adapting billy's filesystem to wazero's [experimental sys interface](https://pkg.go.dev/github.com/tetratelabs/wazero@v1.11.0/experimental/sys). Mostly glue code, except where I had to translate Go errors into POSIX errno values. I had to read both the POSIX spec and the wazero source to figure out how to map errors between the two worlds.

Author: ```suggestion
From there it was a matter of adapting billy's filesystem to wazero's [experimental sys interface](https://pkg.go.dev/github.com/tetratelabs/wazero@v1.11.0/experimental/sys). Mostly glue code, except where I had to translate Go errors into POSIX errno values. I had to read both the POSIX spec, the WASI spec, and the wazero source to figure out how to map errors between the two worlds. I think I'm at least 95% correct, which is likely within the margin of porting error.
```

Adapting that codeinterpreter/python library to the new interface was mostly straightforward, and I ended up with a flow like this:

```go
// from https://tangled.org/xeiaso.net/kefka/blob/main/command/internal/python3/python3.go

func (Impl) Exec(ctx context.Context, ec *command.ExecContext, args []string) error {
	fsConfig := wazero.NewFSConfig().(sysfs.FSConfig).

> 		WithSysFSMount(billyfs.New(ec.FS), "/")

Author: ```suggestion
	fsConfig := wazero.NewFSConfig().
		(sysfs.FSConfig).
		WithSysFSMount(billyfs.New(ec.FS), "/")
```

> func (Impl) Exec(ctx context.Context, ec *command.ExecContext, args []string) error {
> 	fsConfig := wazero.NewFSConfig().(sysfs.FSConfig).
> 		WithSysFSMount(billyfs.New(ec.FS), "/")
> 
> 	config := wazero.NewModuleConfig().
> 		WithStdin(ec.Stdin).WithStdout(ec.Stdout).WithStderr(ec.Stderr).
> 		WithArgs(append([]string{"python3"}, args...)...).
> 		WithName("python3").
> 		WithFSConfig(fsConfig).
> 		WithSysNanosleep().WithSysNanotime().WithSysWalltime()

Author: ```suggestion
	config := wazero.NewModuleConfig().
		// Pipe ExecContext stdio
		WithStdin(ec.Stdin).WithStdout(ec.Stdout).WithStderr(ec.Stderr).
		// Pipe argv
		WithArgs(append([]string{"python3"}, args...)...).
		WithName("python3").
		// Pipe filesystem
		WithFSConfig(fsConfig).
		// Pipe system time
		WithSysNanosleep().WithSysNanotime().WithSysWalltime()
```

	mod, err := runtime.InstantiateModule(ctx, compiled, config)
	if err != nil {

>     // Fit the square peg into the round hole

Author: ```suggestion
        // Fit the square peg into the round hole
```

		if exitErr, ok := errors.AsType[*wsys.ExitError](err); ok {
			if code := exitErr.ExitCode(); code != 0 {
				return interp.ExitStatus(uint8(code))
			}
			return nil
		}
		return err
	}
	return mod.Close(ctx)
}
```

<Conv name="Mara" mood="aha">

>   See? The dependencies such as stdin, stdout, and stderr get _injected_ into

Author: ```suggestion
  See? The dependencies such as `stdin`, `stdout`, and `stderr` get _injected_ into
```

### Using filesystems that don't exist

OK, that handles filesystems that (arguably) exist, like the btrfs volume on my dev box. What about filesystems that don't? For the sake of argument, let's say you want this fake shell to interact with object storage as its main filesystem. At some level all you need to do is adapt the billy interface to object storage using something like [storage-go](https://pkg.go.dev/github.com/tigrisdata/storage-go).

<Conv name="Cadey" mood="coffee">
  Disclaimer, I work at Tigris and developed this library for them. It's
  basically the S3 client with more methods to handle additional Tigris features
  like forks and snapshots. I'll be writing more about it soon.
</Conv>

> After finding [a basic implementation of an S3 -> Billy adapter](https://github.com/a-poor/s3fs), I vendored it into the Kefka repo and swapped out the "real" filesystem in cmd/kefka for an s3fs implementation pointed at a sample Tigris bucket. From there it was down to an iterative process of running commands, finding feature gaps when errors showed up, implementing them, and making sure things work the same against Tigris as they do against a local filesystem.

Author: ```suggestion
After finding [a basic implementation of an S3 -> Billy adapter](https://github.com/a-poor/s3fs), I vendored it into the Kefka repo and swapped out the "real" filesystem in cmd/kefka for an s3fs implementation pointed at a sample Tigris bucket. From there it was down to an iterative process of running commands, finding feature gaps when errors showed up, implementing them, fuzzing, and making sure things work mostly the same against Tigris as they do against a local filesystem.
```

> WASI is cursed: it has no process-level "current working directory," which most programs assume exists. You patch around it by passing a `CWD` envvar, or — easier — just use absolute paths. I haven't hit anything broken in casual use, but expect rough edges. Here be dragons and this code may be known by the state of California to cause cancer.

Author: ```suggestion
WASI is cursed: it has no process-level "current working directory," which most programs assume exists. You patch around it by passing a `CWD` envvar, or just use absolute paths. I haven't hit anything broken in casual use, but expect rough edges. Here be dragons and this code may be known by the state of California to cause cancer.
```

    // implementation, interp falls back to its default ExecHandler that
    // executes the command as a subprocess. This is almost certainly
    // not what you want.
    interp.ExecHandlers(constrainToRegistry(commands)),
    // Wire up per-command pwd state to the filesystem implementation
    interp.CallHandler(billysh.CallHandler(commands, fsys, stdout, stderr)),
    // Handle shell-level filesystem I/O (redirects, glob expansion, etc)
    interp.StatHandler(billysh.FsysStatHandler(commands, fsys)),
    interp.FsysOpenHandler(billysh.FsysOpenHandler(commands, fsys)),
    interp.ReadDirHandler2(billysh.FsysReadDirHandler(commands, fsys)),
  )

  // Read shell commands
  parser := syntax.NewParser()
  fmt.Fprintf(stdout, "$ ")

  // Split input into commands
  for stmts, err := range parser.InteractiveSeq(stdin) {
    if err != nil {
      return err
    }

    if parser.Incomplete() {
      fmt.Fprintf(stdout, "> ")
      continue
    }

    for _, stmt := range stmts {
      err := sh.Run(ctx, stmt)
      if sh.Exited() {
        return err
      }
    }

    // Show prompt
    fmt.Fprintf(stdout, "$ ")
  }

  return nil
}
```

> [The real handler](https://tangled.org/xeiaso.net/kefka/blob/main/cmd/sophia/main.go#L87) is much messier because Python's REPL needs careful buffering, Ctrl-C has to actually cancel things, pty wiring is its own swamp. None of that shows up if it's working. Tab completion and readline polish are easy enough; I'll let you wire those up as an exercise for the reader.

Author: ```suggestion
[The real handler](https://tangled.org/xeiaso.net/kefka/blob/main/cmd/sophia/main.go#L87) is much messier because Python's REPL needs careful buffering, Ctrl-C has to actually cancel things, and [pty](https://en.wikipedia.org/wiki/Pseudoterminal) wiring is its own can of cans of worms. None of that shows up if it's working. Tab completion and readline polish are easy enough; I'll let you wire those up as an exercise for the reader.
```

If you want to try it today, you can ssh into `sophia.xeiaso.net` on port 2222:

```
$ ssh sophia.xeiaso.net -p 2222

> ```

Author: ```suggestion
If you want to try it today, you can ssh into `sophia.xeiaso.net`:

    $ ssh sophia.xeiaso.net
```

You'll get an isolated sandbox in your own [bucket fork/branch](https://www.tigrisdata.com/docs/forks/). Every `ls` is a `ListObjectsV2` against the bucket. Every `qjs` or `python3` runs WebAssembly on the server, wired to that same bucket.

```
$ cat ./samples/hello.js
console.log("Hello, world!");
$ qjs ./samples/hello.js
Hello, world!
```

The demo bucket is seeded with examples. You'll probably have to poke around to find everything. Worst case, run `help`.

I want more experimental WebAssembly hacks like this to exist. I'll keep poking at it.

## Put your programs in clown jail

This has legs. [yeet](https://github.com/TecharoHQ/yeet) could run [Anubis](https://anubis.techaro.lol) builds on Windows; and if management ever makes you babysit AI agents, clown jail is a decent answer.

> Code lives [on Tangled](https://tangled.org/xeiaso.net/kefka). I'm wiring it into [an agent harness](https://github.com/tigrisdata-community/mithras) so I can automate small tools against a local model (Qwen3-30B-A3B right now, good enough for what I'm doing).

Author: ```suggestion
The code lives [on Tangled](https://tangled.org/xeiaso.net/kefka). I'm wiring it into [an agent harness](https://github.com/tigrisdata-community/mithras) so I can automate small tools against a local model (I'm loving Qwen3-36B-A3B).
```

> There's a sister post on the Tigris blog that goes deeper into the AI-agent angle and the porting work using Claude Code. If you want, you can check it out here: [Give your agents disposable environments in Go](https://tigrisdata.com/blog/agent-sandbox-go)

Author: Do I social card this? I may want to.
